File: frontend/src/lib/history.ts
import type { HistoryEntry, StreamingDelta } from "@/types";

export function historyTimestampToMs(timestamp: number): number {
  return timestamp > 1_000_000_000_000
    ? timestamp
    : Math.round(timestamp * 1000);
}

export function reduceDeltas(deltas: StreamingDelta[]) {
  let content = "";
  let thinking = "";
  const toolResults = new Map<string, string>();
  const sentMessages = new Map<string, { toIds: string[]; text: string }>();
  const receivedMessages = new Map<string, { fromId: string; text: string }>();
  const messageOrder: Array<{ kind: "sent" | "received"; messageId: string }> =
    [];

  for (const delta of deltas) {
    switch (delta.type) {
      case "ContentDelta":
        content += delta.text;
        break;
      case "ThinkingDelta":
        thinking += delta.text;
        break;
      case "ToolResultDelta":
        toolResults.set(
          delta.tool_call_id,
          (toolResults.get(delta.tool_call_id) ?? "") + delta.text,
        );
        break;
      case "SentMessageDelta":
        if (!sentMessages.has(delta.message_id)) {
          messageOrder.push({ kind: "sent", messageId: delta.message_id });
        }
        sentMessages.set(delta.message_id, {
          toIds: delta.to_ids,
          text: (sentMessages.get(delta.message_id)?.text ?? "") + delta.text,
        });
        break;
      case "ReceivedMessageDelta":
        if (!receivedMessages.has(delta.message_id)) {
          messageOrder.push({ kind: "received", messageId: delta.message_id });
        }
        receivedMessages.set(delta.message_id, {
          fromId: delta.from_id,
          text:
            (receivedMessages.get(delta.message_id)?.text ?? "") + delta.text,
        });
        break;
    }
  }

  return {
    content,
    thinking,
    toolResults,
    sentMessages,
    receivedMessages,
    messageOrder,
  };
}

function serializeHistoryValue(value: unknown): string {
  if (value === null || value === undefined) {
    return "";
  }

  if (typeof value === "string") {
    return value;
  }

  try {
    return JSON.stringify(value);
  } catch {
    return String(value);
  }
}

function getHistoryEntryDedupKey(entry: HistoryEntry): string {
  const timestamp = historyTimestampToMs(entry.timestamp);

  switch (entry.type) {
    case "ReceivedMessage":
      if (entry.message_id) {
        return `${entry.type}:${entry.message_id}`;
      }
      return [
        entry.type,
        timestamp,
        entry.from_id ?? "",
        entry.content ?? "",
      ].join(":");
    case "SentMessage":
      if (entry.message_id) {
        return `${entry.type}:${entry.message_id}`;
      }
      return [
        entry.type,
        timestamp,
        (entry.to_ids ?? []).join(","),
        entry.content ?? "",
      ].join(":");
    case "ToolCall":
      if (entry.tool_call_id) {
        return `${entry.type}:${entry.tool_call_id}`;
      }
      return [
        entry.type,
        timestamp,
        entry.tool_name ?? "",
        serializeHistoryValue(entry.arguments),
        entry.result ?? "",
        entry.streaming ? "streaming" : "final",
      ].join(":");
    case "StateEntry":
      return [
        entry.type,
        timestamp,
        entry.state ?? "",
        entry.reason ?? "",
      ].join(":");
    default:
      return [entry.type, timestamp, entry.content ?? ""].join(":");
  }
}

function shouldReplaceDuplicateHistoryEntry(
  existing: HistoryEntry,
  candidate: HistoryEntry,
): boolean {
  if (existing.type !== "ToolCall" || candidate.type !== "ToolCall") {
    return false;
  }

  if (existing.streaming && !candidate.streaming) {
    return true;
  }

  return !existing.result && Boolean(candidate.result);
}

function dedupeHistoryEntries(entries: HistoryEntry[]): HistoryEntry[] {
  const deduped: HistoryEntry[] = [];
  const seenIndexes = new Map<string, number>();

  for (const entry of entries) {
    const key = getHistoryEntryDedupKey(entry);
    const existingIndex = seenIndexes.get(key);

    if (existingIndex === undefined) {
      seenIndexes.set(key, deduped.length);
      deduped.push(entry);
      continue;
    }

    const existing = deduped[existingIndex];
    if (shouldReplaceDuplicateHistoryEntry(existing, entry)) {
      deduped[existingIndex] = entry;
    }
  }

  return deduped;
}

export function mergeHistoryWithDeltas({
  history,
  incremental,
  deltas,
  fetchedAt,
}: {
  history: HistoryEntry[];
  incremental?: HistoryEntry[];
  deltas?: StreamingDelta[];
  fetchedAt: number;
}): HistoryEntry[] {
  const base = dedupeHistoryEntries(
    incremental && incremental.length > 0
      ? [...history, ...incremental]
      : [...history],
  );

  if (!deltas || deltas.length === 0) {
    return base;
  }

  const {
    content,
    thinking,
    toolResults,
    sentMessages,
    receivedMessages,
    messageOrder,
  } = reduceDeltas(deltas);
  const now = fetchedAt / 1000;

  if (thinking) {
    base.push({
      type: "AssistantThinking",
      content: thinking,
      timestamp: now,
      streaming: true,
    } satisfies HistoryEntry);
  }

  if (content) {
    base.push({
      type: "AssistantText",
      content,
      timestamp: now,
      streaming: true,
    } satisfies HistoryEntry);
  }

  if (toolResults.size > 0) {
    for (const [toolCallId, resultText] of toolResults) {
      for (let i = base.length - 1; i >= 0; i -= 1) {
        const entry = base[i];
        if (
          entry.type === "ToolCall" &&
          entry.tool_call_id === toolCallId &&
          entry.streaming
        ) {
          base[i] = { ...entry, result: resultText };
          break;
        }
      }
    }
  }

  for (const item of messageOrder) {
    if (item.kind === "sent") {
      const sent = sentMessages.get(item.messageId);
      if (!sent) {
        continue;
      }
      if (
        base.some(
          (entry) =>
            entry.type === "SentMessage" && entry.message_id === item.messageId,
        )
      ) {
        continue;
      }
      base.push({
        type: "SentMessage",
        message_id: item.messageId,
        to_ids: sent.toIds,
        content: sent.text,
        timestamp: now,
        streaming: true,
      } satisfies HistoryEntry);
      continue;
    }

    const received = receivedMessages.get(item.messageId);
    if (!received) {
      continue;
    }
    if (
      base.some(
        (entry) =>
          entry.type === "ReceivedMessage" &&
          entry.message_id === item.messageId,
      )
    ) {
      continue;
    }
    base.push({
      type: "ReceivedMessage",
      message_id: item.messageId,
      from_id: received.fromId,
      content: received.text,
      timestamp: now,
      streaming: true,
    } satisfies HistoryEntry);
  }

  return base;
}

File: app/models/history.py
from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any

from app.models.base import Serializable
from app.models.content import (
    ContentPart,
    content_parts_to_text,
    deserialize_content_parts,
)


@dataclass
class SystemEntry(Serializable):
    content: str
    timestamp: float = field(default_factory=time.time)


@dataclass
class ReceivedMessage(Serializable):
    from_id: str
    parts: list[ContentPart] = field(default_factory=list)
    content: str = ""
    message_id: str | None = None
    timestamp: float = field(default_factory=time.time)

    def __post_init__(self) -> None:
        if self.parts and not self.content:
            self.content = content_parts_to_text(self.parts)
        elif self.content and not self.parts:
            self.parts = deserialize_content_parts(None, fallback_text=self.content)


@dataclass
class AssistantText(Serializable):
    parts: list[ContentPart] = field(default_factory=list)
    content: str = ""
    timestamp: float = field(default_factory=time.time)

    def __post_init__(self) -> None:
        if self.parts and not self.content:
            self.content = content_parts_to_text(self.parts)
        elif self.content and not self.parts:
            self.parts = deserialize_content_parts(None, fallback_text=self.content)


@dataclass
class SentMessage(Serializable):
    to_id: str
    parts: list[ContentPart] = field(default_factory=list)
    content: str = ""
    message_id: str | None = None
    timestamp: float = field(default_factory=time.time)

    def __post_init__(self) -> None:
        if self.parts and not self.content:
            self.content = content_parts_to_text(self.parts)
        elif self.content and not self.parts:
            self.parts = deserialize_content_parts(None, fallback_text=self.content)


@dataclass
class AssistantThinking(Serializable):
    content: str
    timestamp: float = field(default_factory=time.time)


@dataclass
class StateEntry(Serializable):
    state: str
    reason: str = ""
    timestamp: float = field(default_factory=time.time)


@dataclass
class ToolCall(Serializable):
    tool_name: str
    tool_call_id: str
    arguments: dict[str, Any]
    result: str | None = None
    streaming: bool = False
    timestamp: float = field(default_factory=time.time)


@dataclass
class ErrorEntry(Serializable):
    content: str
    timestamp: float = field(default_factory=time.time)


@dataclass
class CommandResultEntry(Serializable):
    command_name: str
    content: str
    include_in_context: bool = False
    timestamp: float = field(default_factory=time.time)


HistoryEntry = (
    SystemEntry
    | ReceivedMessage
    | AssistantText
    | SentMessage
    | AssistantThinking
    | StateEntry
    | ToolCall
    | ErrorEntry
    | CommandResultEntry
)


def deserialize_history_entry(data: dict[str, Any]) -> HistoryEntry:
    entry_type = data.get("type")
    timestamp = data.get("timestamp")
    timestamp_value = timestamp if isinstance(timestamp, (int, float)) else time.time()

    if entry_type == "SystemEntry":
        return SystemEntry(
            content=str(data.get("content", "")),
            timestamp=timestamp_value,
        )
    if entry_type == "ReceivedMessage":
        parts = deserialize_content_parts(
            data.get("parts"),
            fallback_text=str(data.get("content", "")),
        )
        return ReceivedMessage(
            from_id=str(data.get("from_id", "")),
            parts=parts,
            content=content_parts_to_text(parts),
            message_id=(
                str(data["message_id"])
                if isinstance(data.get("message_id"), str)
                else None
            ),
            timestamp=timestamp_value,
        )
    if entry_type == "AssistantText":
        parts = deserialize_content_parts(
            data.get("parts"),
            fallback_text=str(data.get("content", "")),
        )
        return AssistantText(
            parts=parts,
            content=content_parts_to_text(parts),
            timestamp=timestamp_value,
        )
    if entry_type == "SentMessage":
        parts = deserialize_content_parts(
            data.get("parts"),
            fallback_text=str(data.get("content", "")),
        )
        to_id = (
            str(data["to_id"])
            if isinstance(data.get("to_id"), str)
            else (
                str(data["to_ids"][0])
                if isinstance(data.get("to_ids"), list)
                and data["to_ids"]
                and isinstance(data["to_ids"][0], str)
                else ""
            )
        )
        return SentMessage(
            to_id=to_id,
            parts=parts,
            content=content_parts_to_text(parts),
            message_id=(
                str(data["message_id"])
                if isinstance(data.get("message_id"), str)
                else None
            ),
            timestamp=timestamp_value,
        )
    if entry_type == "AssistantThinking":
        return AssistantThinking(
            content=str(data.get("content", "")),
            timestamp=timestamp_value,
        )
    if entry_type == "StateEntry":
        return StateEntry(
            state=str(data.get("state", "")),
            reason=str(data.get("reason", "")),
            timestamp=timestamp_value,
        )
    if entry_type == "ToolCall":
        arguments = data.get("arguments")
        return ToolCall(
            tool_name=str(data.get("tool_name", "")),
            tool_call_id=str(data.get("tool_call_id", "")),
            arguments=arguments if isinstance(arguments, dict) else {},
            result=str(data["result"]) if isinstance(data.get("result"), str) else None,
            streaming=bool(data.get("streaming", False)),
            timestamp=timestamp_value,
        )
    if entry_type == "ErrorEntry":
        return ErrorEntry(
            content=str(data.get("content", "")),
            timestamp=timestamp_value,
        )
    if entry_type == "CommandResultEntry":
        return CommandResultEntry(
            command_name=str(data.get("command_name", "")),
            content=str(data.get("content", "")),
            include_in_context=bool(data.get("include_in_context", False)),
            timestamp=timestamp_value,
        )
    raise ValueError(f"Unknown history entry type: {entry_type}")


def deserialize_history_entries(items: list[dict[str, Any]]) -> list[HistoryEntry]:
    return [deserialize_history_entry(item) for item in items]
